### ISSUE TITLE
[Feature Request] MultiBlocBuilder

### ISSUE THREAD

**Issue by pczn0327** [felangel/bloc#538]
Sometimes I would need to access different blocs in the same widget, and the code goes pretty messy. I am thinking maybe we can have something like `MultiBlocBuilder` which serve the similar purpose of `MultiRepositoryProvider` and `MultiBlocProvider`.

**Comment by felangel** [comment]
Hi @pczn0327 👋 
Thanks for opening an issue!

This has been brought up in the past but we have decided against it for [several reasons](https://github.com/felangel/bloc/issues/428).

Can you please provide some use-cases? If there are valid use-cases and enough people would benefit from it, then I'm open to including it. Thanks 👍

**Comment by pczn0327** [comment]
Hi @felangel , I can't tell you the exact use case, but I've simulated similar use case in [this repo](https://github.com/pczn0327/MultiBlocBuilder-Use-Case).

Use case usage comes from different form is rendered based on input from user. 

I hope that helps.

**Comment by henriquearthur** [comment]
I'm facing the same problem on my project. I believe only apps from medium-large complexity faces this.

From @pczn0327 example:

```
          Expanded(
            child: BlocBuilder<BlueBloc, bool>(
              bloc: blueBloc,
              builder: (context, blueState) {
                return BlocBuilder<GreenBloc, bool>(
                  bloc: greenBloc,
                  builder: (context, greenState) {
                    return BlocBuilder<RedBloc, bool>(
                      bloc: redBloc,
                      builder: (context, redState) {
```

This can get worse when we have to do a few `if`s to check for current state. A `MultiBlocBuilder` would make the code more readable.

In the meantime, or in case `MultiBlocBuilder`  doesn't get implemented, do you recommend any approach when using multile blocs like this @felangel?

**Comment by felangel** [comment]
@henriquearthur yup in the meantime you need to nest BlocBuilders just as you've shown 👍 

I'm putting together an implementation that will expose BlocBuilder2-BlocBuilder9 which can be used like:

```dart
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Example')),
      body: BlocBuilder2<BlueBloc, BlueState, GreenBloc, GreenState>(
        builder: (context, blueState, greenState) { ... },
      ),
    );
  }
```

Please let me know what you think of the proposed api 🙏

**Comment by pczn0327** [comment]
Hi @felangel , does that means if I need to use 5 blocs, I will need to do as follow?
`BlocBuilder5<BlueBloc, BlueState, GreenBloc, GreenState, RedBloc, RedState, WhiteBloc, WhiteState, BlackBloc, BlackState>`?

**Comment by felangel** [comment]
@pczn0327 yup or alternatively you will need to manually provide the blocs like:
```dart
BlocBuilder5(
  blocA: blueBloc,
  blocB: greenBloc,
  blocC: redBloc,
  blocD: whiteBloc,
  blocE: blackBloc,
  builder: (context, blueState, greenState, redState, whiteState, blackState) { ... }
)
```

Without that Dart will not be able to infer the state types and all states will resolve to type `dynamic`.

**Comment by pczn0327** [comment]
Cool! Sounds like a good start for it! Looking forward for the solution!

**Comment by adityadroid** [comment]
This will definitely be useful. Infact the first time I came across a usecase where I had to nest multiple BlocBuilders, I went through the entire documentation of bloc to see if there was a cleaner way of doing it. Nesting multiple BlocBuilder Widgets somehow felt wrong to me.

**Comment by ThinkDigitalSoftware** [comment]
I think 9 is too many. At that point code should be refactored. If course we have to account for edge cases, but in that case, there's no end.

**Comment by andredsnogueira** [comment]
It would be awesome.

**Comment by felangel** [comment]
> I think 9 is too many. At that point code should be refactored. If course we have to account for edge cases, but in that case, there's no end. Imagine what the code for that widget is going to look like!

I completely agree. I was just messing around and I'm still not sold on if this belongs in `flutter_bloc` because it's not really a "core" component. It's more of a convenience thing which might be better as an independent package.

**Comment by warriorCoder** [comment]
I'm with Felix on this one. It seems like this isn't part of the core functionality, and I'd also like to see a real live use case. Why would you need to nest BlocBuilders in a real app?

I'm wondering if there isn't a different way to build to the widget (like maybe it is trying to do too many things, or the blocs are too granular)?

**Comment by zaralockheart** [comment]
I actually faced this issue in my current project. My colleague asked "I have 3 blocs and they are depends on each other, is it okay to make multiple bloc builder?" I told him, rather than doing that, maybe put them all in one bloc, at least those business logic is placed in a room not different room.

IMO, this idea is good, but maybe if there a need to build based on multiple bloc, I think its better to check if there is a way to refactor the code first.

**Comment by ThinkDigitalSoftware** [comment]
You can also combine your streams ahead of time and provide the combined stream to a single bloc. This will simplify your build methods.

**Comment by ThinkDigitalSoftware** [comment]
I believe that they can also separate the widgets by layers. For example, if they require the authentication state in a widget 5 layers down, that should be part of the state that's passed from above .  I agree that their blocs may be too granular

**Comment by DeepStyles** [comment]
In this feature, I wouldn't want to rebuild unnecessary having consumed multiple blocs... Provider package has updated with selector widget recently which only rebuild when related Bloc changes. (My widget go crazy building, if one f the multiple blocs provided is related to animation)

**Comment by felangel** [comment]
@DeepStyles you'd still be able to provide conditions for each bloc like:
```dart
@override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Example')),
      body: BlocBuilder2<BlueBloc, BlueState, GreenBloc, GreenState>(
        conditionA: (prevBlueState, nextBlueState) => ...,
        conditionB: (prevGreenState, nextGreenState) => ..., 
        builder: (context, blueState, greenState) { ... },
      ),
    );
  }
```

Either way I'm leaning towards not moving forward with this feature because I don't see many use-cases and nested `BlocBuilders` are more readable/maintainable in my opinion. Also, as many people have pointed out if you have lots of nested `BlocBuilders` there's a good chance you need to refactor your blocs.

**Comment by felangel** [comment]
@henriquearthur I took a closer look at your use-case and I would argue that you can refactor to just have a `ColorBloc` rather than a `RedBloc`, `GreenBloc`, and `BlueBloc`. You can have a `ColorChanged` event which takes `r`, `g`, and `b` values and a `ColorState` which keeps track of the current `r`, `g`, and `b` value. I don't think you would need a `MultiBlocBuilder` anymore at that point.

**Comment by felangel** [comment]
Closing this for now because I have not seen a compelling use case for this. If anyone feels like they have a good use case please let me know and I'm happy to re open it 👍

**Comment by ThinkDigitalSoftware** [comment]
@felangel nitpicking, but your label won'tfix conveys that this topic is closed for discussion

**Comment by felangel** [comment]
@ThinkDigitalSoftware sorry! I removed won't fix and re-added discussion/feedback wanted 👍

**Comment by MichelMelhem** [comment]
In the case of a complex view that use multiple source of data,
I think that the widget you mentionned above would be very useful.

**Comment by felangel** [comment]
@MichelMelhem can you give an example?

**Comment by algirdasmac** [comment]
Hello,

I was watching this feature request and I think I have one use case, where this MultiBlocBuilder would be usefull. Lets look at this example:
```dart
class MultiBlocExampleWidget extends StatefulWidget {
  const MultiBlocExampleWidget({Key key}) : super(key: key);

  @override
  _MultiBlocExampleWidgetState createState() => _MultiBlocExampleWidgetState();
}

class _MultiBlocExampleWidgetState extends State<MultiBlocExampleWidget> {
  final int totalSteps = 5;
  int selectedStep = 0;

  @override
  Widget build(BuildContext context) {
    return Stepper(
      currentStep: selectedStep,
      onStepCancel: () {
        if (selectedStep > 0) {
          setState(() => selectedStep = selectedStep - 1);
        }
      },
      onStepContinue: () {
        if (selectedStep < totalSteps - 1) {
          setState(() => selectedStep = selectedStep + 1);
        }
      },
      onStepTapped: (index) => setState(() => selectedStep = index),
      steps: [
        Step(
          title: Text('Address'),
          subtitle: Text('selectedAddress from AddressBloc'),
          content: Text('address list from AddressBloc'),
        ),
        Step(
          title: Text('Delivery time'),
          subtitle: Text('selectedDeliveryOption from DeliveryBloc'),
          content: Text('delivery options from DeliveryBloc'),
        ),
        Step(
          title: Text('Payment method'),
          subtitle: Text('selectedPaymentMethod from PaymentBloc'),
          content: Text('payment options from PaymentBloc'),
        ),
        Step(
          title: Text('Phone number'),
          subtitle: Text('phoneNumber from UserBloc'),
          content: Text('edit phone number with UserBloc'),
        ),
        Step(
          title: Text('Confirm order'),
          subtitle: Text('orderTotal from OrderBloc'),
          content: Text('order totals info'),
        ),
      ],
    );
  }
}
```

So I would like to have each step have its own BLoC:
1. Selecting your delivery address (address bloc)
2. Selecting delivery time (delivery bloc)
3. Selecting payment type (payment bloc)
4. Selecting phone number (user bloc)
5. Confirming order (order bloc)

So in my case I have to nest all BlocBuilders above Stepper widget, just because `Stepper` widget parameter `steps` accepts only `List<Step>`. Same goes for `ExpansionPanelList` widget and I think there is more.
Also these five blocs merged into one big BLoC would be a real mess (tried that), because they are used also in other parts of the app, just for their own functionality.

This is my real world example that I had to deal with - using BlocBuilder in Flutter widgets, where array of widgets can only be of specific type (not `List<Widget>`).

**Comment by ThinkDigitalSoftware** [comment]
@algirdasmac but why would you have separate blocs for these? Why not have a single `OrdersBloc`? Think of a bloc like a manager in a business. Would you have a Manager over getting the customers address and a separate manager for their payment type? Not really. Your blocs can handle multiple events if they extend the same class, so use that to your advantage and keep things simple. Take a look at Redux examples or other examples with this library that yield states using currentState.copyWith functions to  yield new States without losing old data. That way you can make a single block that just focuses on managing your orders page

**Comment by algirdasmac** [comment]
@ThinkDigitalSoftware I separated each Bloc, so I would not have one Bloc with massive amount of Events and State object would be manageable. Look at it this way:
1. Payments Bloc issues you a new payment card, removes one if you need, selects your preferred payment method, syncs all your stored payment methods and processes payment.
2. Delivery and Order bloc is actually the same, I just wanted to make my example more dramatic :). Still this bloc is responsible of syncing order info with database, selecting available delivery times, save additional comments.
3. Address Bloc writes, edits, deletes your addresses, syncs address list with database, also searches for address in google places.
4. User Bloc is the one, where you set your email, phone number, password, name and etc.

Would you have one manager in the company to issue cards, select delivery time for your order and search for available addresses? Probably not. Each of the bloc have its own page (Payments page, Addresses page, Orders page, User profile page), but this one checkout page brings everyone together. I tried doing everything in one BLoC, but then the code becomes really messy and hard to manage - state object is massive - list of available addresses, selected address, list of payment methods, selected payment method, list of delivery times, selected delivery time, all the user information and etc. Also events are repeated as in smaller blocs, and I needed to copy/maintain same code in two places.

I hope this answers your question.

**Comment by ThinkDigitalSoftware** [comment]
Thank can be understandable, but if you had to create an order, would you call all of your managers over to give you an individual piece of that the data to make a single order? It would be more likely that this information was already compiled when you got to this section as a single Order object, which would be compiled by the OrdersBloc. It could be ok to keep this many blocs, but have a shared state or object that you can access from your blocs that would total an `Order` object so your widget isn't doing the business logic (like assembling the Order). Because anytime you need to access this order, you're going to need to build this same logic again and it'll have to be in the widget tree since that's where you'll have access to the context in order to look up your blocs.
It sounded like, from your description of the state object that you had less than 10 fields. Your 6 lists, your `User` object, etc. If you saw Redux state classes you'd die. This is small compared to what it can get to in Redux. They believe in a single source of truth. One state for the entire app, broken up into smaller classes, etc. So its not an unhealthy design pattern if it's organized properly. But creating more blocs than you need can lead to complication like this. It's up to you, of course. Just wanted to present an alternate way to look at it

**Comment by algirdasmac** [comment]
There is probably no right way - each has its own 😄. To be honest, having smaller blocs is really easy to manage, especially when we reuse our logic in web and flutter. Three full time developers are working on BLoC's and it is working fine for us yet, but I am really thankful for your insights and suggestions!

What I tried to do is to show an example, where MultiBlocBuilder would be useful in flutter widget library. Maybe my use case is not the best, but if you need to use two or more blocs for Stepper/ExpansionPanelList it becomes really messy. But this is very rare use-case.

**Comment by ThinkDigitalSoftware** [comment]
I hope you find a working solution!

**Comment by amreniouinnovent** [comment]
@felangel I think we need a better sample for repository builder so we avoid nested bloc builders

**Comment by tenhobi** [comment]
It would be cool to have some tips how to do complicated things to avoid this problem, I would like to have something to help me improve it. <3

**Comment by ThinkDigitalSoftware** [comment]
@tenhobi what I've done previously is use a combineLatest function from RxDart to merge my streams and output a stream with a single object that is updated anytime any of the other streams are updated and then subscribe to that using a StreamBuilder

**Comment by ThinkDigitalSoftware** [comment]
@felangel just came across a use case. In my case, the devs have decided to have a separate bloc per news list tab, which is separated by region. They've added a tab which should show news from all of them so now I need to combine the outputs. This could be tricky with the current nested blocs and conditions. I haven't started reasoning on a solution yet, but this is the first thing I thought of

**Comment by RollyPeres** [comment]
@ThinkDigitalSoftware that sounds like you should use a single bloc, filter your data on those tabs and show everything in your latest added tab.

**Comment by felangel** [comment]
@ThinkDigitalSoftware yeah as @RollyPeres mentioned, have you considered having a single news bloc which you could filter by region or is that out of the question? Would love to learn more 😄

**Comment by ThinkDigitalSoftware** [comment]
> @ThinkDigitalSoftware that sounds like you should use a single bloc, filter your data on those tabs and show everything in your latest added tab.

That's possible. My issue is that the tech lead doesn't want me spending too much time on what would be considered refinements at the moment.

Here's a question that should be answered though.

When does a library focus on best practices vs freedom of style? 
This is why pendantic and effective_dart are simply linting suggestions rather than enforced rules. and why `var` still exists, etc. We should account for the edge cases rather than only adding something when someone can prove it's impossible or less favorable to implement another way.
We have a lot of work ahead of us if we're going to police the programming styles of our libraries' users.

I ended up solving this problem by using a CombineLatestStream. I know it's a sin, but it was WAY faster than consolidating 3-4 blocs into one.

**Comment by felangel** [comment]
Thanks for sharing and for the feedback! I’m a bit curious why that feature was considered a refinement.

> We should account for the edge cases rather than only adding something when someone can prove it's impossible or less favorable to implement another way.

Can you elaborate a bit on what you mean by this?

**Comment by ThinkDigitalSoftware** [comment]
> Thanks for sharing and for the feedback! I’m a bit curious why that feature was considered a refinement.

I didn't ask, but based on other things I took the time to fix and was counseled on, this is a bigger task.



> > We should account for the edge cases rather than only adding something when someone can prove it's impossible or less favorable to implement another way.
> 
> Can you elaborate a bit on what you mean by this?

From the viewpoint of a new dev, even if combining blocs isn't a favorable practice, I don't believe it's your business to decide that. Rather than ask why add it, are we able to answer why not? If we have a `MultiBlocListener`, it seems natural to have its companion that can build. We know it's possible and that there are use cases to listening to multiple blocs. Why **not** add a builder?

**Comment by felangel** [comment]
> From the viewpoint of a new dev, even if combining blocs isn't a favorable practice, I don't believe it's your business to decide that. Rather than ask why add it, are we able to answer why not? If we have a MultiBlocListener, it seems natural to have its companion that can build. We know it's possible and that there are use cases to listening to multiple blocs. Why not add a builder?

@ThinkDigitalSoftware fair enough. I'll revisit this request this weekend 👍

**Comment by RollyPeres** [comment]
@ThinkDigitalSoftware I'm not ruling out the need of combining blocs sometimes, but in my experience, whenever there's a need to build data driven UI's based on more than 2 blocs, it might be a side effect of not properly designing the blocs and their responsibilities.

> In my case, the devs have decided to have a separate bloc per news list tab, which is separated by region.

Why not go for a single base bloc which receives a region as a parameter ? This way you guys could have dropped a bloc on top of each tab and have reusable logic. Curious what your thoughts are on this approach @felangel, you endorse it or you consider region parameter as state?  😃

> They've **added a tab** which should show news from all of them **so now I need to** combine the outputs.

This kind of scenarios happen a lot, and it's not bloc's fault. They should be taken into consideration from initial design.

**Comment by ThinkDigitalSoftware** [comment]
> @ThinkDigitalSoftware I'm not ruling out the need of combining blocs sometimes, but in my experience, whenever there's a need to build data driven UI's based on more than 2 blocs, it might be a side effect of not properly designing the blocs and their responsibilities.
> 
> > In my case, the devs have decided to have a separate bloc per news list tab, which is separated by region.
> 
> Why not go for a single base bloc which receives a region as a parameter ? This way you guys could have dropped a bloc on top of each tab and have reusable logic. Curious what your thoughts are on this approach @felangel, you endorse it or you consider region parameter as state?  😃
> 
> > They've **added a tab** which should show news from all of them **so now I need to** combine the outputs.
> 
> This kind of scenarios happen a lot, and it's not bloc's fault. They should be taken into consideration from initial design.

You're right in all your replies, but the fact still remains that I can't touch this section of the logic, so I have to work with what I have. Second, there are many MANY ways in which every dev can improve their logic and code. But why cripple them by not providing options and forcing them to refractor? What benefits do we get? This is why so many devs go with Provider in their state management solution because it **does** provide this option and doesn't restrict them to best practices. Someone may surprise us with a valid use case and then we'll say, "Man, I could have added that in the first place".

Again, I understand and agree with the reasons why it should be left out, but it still doesn't mean that it's the best idea. We're programmers. Flexibility is the name of the game.

**Comment by RollyPeres** [comment]
@ThinkDigitalSoftware I completely agree that flexibility is important.

> What benefits do we get?

From what I've seen so far, bloc is tailored towards enforcing good practices no matter who's writing the code; having someone dropping some fancy opinionated code is harder to do, and that's probably for the best, because it makes for a unified way of writing code. This will make all developers understand it easier and adapt to it way quicker. I believe flexibility comes from how you chose to implement bloc. And each bloc is a `Stream` so sky is the limit what you can achieve with streams ✌

> But why cripple them by not providing options and forcing them to refractor? 

`combineLatest` is a perfectly viable solution if you're trying to ship a quick solution. But I think in long term you get in a way worse spot by not refactoring and that applies to everything code related. If you need to make some workaround to solve an issue, next time you bump into this you'll need to make 3 workarounds, and things will get out of hand rather quickly. This is just a matter of developer perspective I guess(or team lead 😉).

> This is why so many devs go with Provider in their state management solution because it **does** provide this option and doesn't restrict them to best practices.

Are we speaking about the same `provider` here ? because I've rarely seen someone more obsessed with enforcing his own practices than Remi 😂 And he's done a brilliant job doing that and the package benefited from that 🙂 `provider` indeed offers overloads for this kind of situations. But I would still refactor and aim for having at most 2 `BlocBuilder`'s in a widget simply because those overloads would be way too verbose for my taste. If I'd ever need one I'd take it as a sign that I'm probably off the track.

**Comment by ThinkDigitalSoftware** [comment]
Well yes, the same `provider` but then again, the package doesn't aim to solve state management as you know, so it's not concerned with how developers use it outside its scope. I guess that's a good point in favor of keeping it out, and a reason why I'm a die hard fan of the `flutter_bloc` family.

Since I didn't have authorization to spend the time to refactor this properly, flutter_bloc didn't come through for me with a solution to my problem and using combine latest took me many lines more to implement than having it done for me. We could be a lesser used library for the elites or have wider adoption with inclusion and features. The only issue I see is forcing @felangel to make an example for this in the docs without throwing up! 😂

**Comment by RollyPeres** [comment]
> I guess that's a good point in favor of keeping it out, and a reason why I'm a die hard fan of the `flutter_bloc` family.

Bad news for you, `flutter_bloc` uses `provider` under the hood, you can't escape it 😂

> The only issue I see is forcing @felangel to make an example for this in the docs without throwing up! 😂

I wish you the best of luck 😬

**Comment by ThinkDigitalSoftware** [comment]
By keeping it out, I was referring to the `MultiBlocProvider`

**Comment by felangel** [comment]
Sorry for being late to the party haha.

> The only issue I see is forcing @felangel to make an example for this in the docs without throwing up!

@ThinkDigitalSoftware an example of what?

> By keeping it out, I was referring to the MultiBlocProvider

So you’re saying you prefer MultiBlocProvider not be part of flutter_bloc?

**Comment by ThinkDigitalSoftware** [comment]
You usually put good use cases for the widgets you introduce. You're going to have to find one for this, lol

I was saying that it's an argument against including it that sounds reasonable, but there are more on the 'for' side

**Comment by MichelMelhem** [comment]
For me a common use case is that in the same page I need to get some user data (ex: if the user has a valid account) from like an ```Authbloc``` , but in addition of that I have to fetch some other data (ex: a list of create todos).
In this case it would be useful to have this bloc.
But in fairness i think that this type of cases are pretty mandatory and do not need a special widget.

But what I think would be more useful to still have the classic bloc builder and in addition of that have as a second ootiobal argument a possibility to pass an optional blocs so it would not break actual code . An example of my thoughts :
```
Blocbuilder<bloc1, stateofbloc1><bloc2,stateofbloc2>(
context: context,
builder:(context, bloc1, stateofbloc1, bloc2,stateofbloc2)
{return ...}

)
```

Voilà !

**Comment by ThinkDigitalSoftware** [comment]
I don't believe you can optionally place generics like that. It would have to be a separate class built for handling 4 generic types, which would be what we were discussing

On May 7, 2020 7:10 AM, Michel M <notifications@github.com> wrote:

For me a common use case is that in the same page I need to get some user data (ex: if the user has a valid account) from like an Authbloc , but in addition of that I have to fetch some other data (ex: a list of create todos).
In this case it would be useful to have this bloc.
But in fairness i think that this type of cases are pretty mandatory and do not need a special widget.

But what I think would be more useful to still have the classic bloc builder and in addition of that have as a second ootiobal argument a possibility to pass an optional blocs so it would not break actual code . An example of my thoughts :

Blocbuilder<bloc1, stateofbloc1><bloc2,stateofbloc2>(

context: context,

builder:(context, bloc1, stateofbloc1, bloc2,stateofbloc2)

{return ...}



)



Voilà !

**Comment by Allan-Nava** [comment]
> For me a common use case is that in the same page I need to get some user data (ex: if the user has a valid account) from like an `Authbloc` , but in addition of that I have to fetch some other data (ex: a list of create todos).
> In this case it would be useful to have this bloc.
> But in fairness i think that this type of cases are pretty mandatory and do not need a special widget.
> 
> But what I think would be more useful to still have the classic bloc builder and in addition of that have a second optional argument that give the possibility to pass an optional bloc so it would not break actual code .
> An example of my thoughts :
> 
> ```
> Blocbuilder<bloc1, stateofbloc1><bloc2,stateofbloc2>(
> context: context,
> builder:(context, bloc1, stateofbloc1, bloc2,stateofbloc2)
> {return ...}
> 
> )
> ```
> 
> Voilà !
> 
> PS: I know this feature request is quite old but as in some niche cases this feature could be useful and I think my solution could make everyone happy without over-ingeniring the lib

How can I implement this feature? I need to handle two BLoCs one for LiveEvents and one for Categories

**Comment by felangel** [comment]
@Allan-Nava as of now you can nest the BlocBuilders.

**Comment by Allan-Nava** [comment]
> @Allan-Nava as of now you can nest the BlocBuilders.

Ok, but the UI is:
![photo_2020-05-20_18-55-35](https://user-images.githubusercontent.com/22498435/82565883-6c640b80-9b7b-11ea-8cc2-d4064f9febb3.jpg)

I have to fill the dropdown with categories

**Comment by felangel** [comment]
What is the issue? You can have a categories bloc which manages the state of the categories which is used by the dropdown section via a BlocBuilder.

**Comment by Allan-Nava** [comment]
> What is the issue? You can have a categories bloc which manages the state of the categories which is used by the dropdown section via a BlocBuilder.

Right now the memory leak, and i need to show in the same time cause I do two http requests.
Maybe this is the solution is:

> you should probably create a single bloc for that feature and the bloc can handle making both network requests. If they can be made in parallel you can use Future.wait([...]). Then in your UI you just need one BlocBuilder.

**Comment by Allan-Nava** [comment]
> > What is the issue? You can have a categories bloc which manages the state of the categories which is used by the dropdown section via a BlocBuilder.
> 
> Right now the memory leak, and i need to show in the same time cause I do two http requests.
> Maybe this is the solution is:
> 
> > you should probably create a single bloc for that feature and the bloc can handle making both network requests. If they can be made in parallel you can use Future.wait([...]). Then in your UI you just need one BlocBuilder.

I tried but I got the same problem and the app crash.

**Comment by rafaelmpessoa** [comment]
@felangel thanks for your support lib and package, which helps me a lot.
I have one case, I have a widget `ContinueButton` which has two properties `isLoading` and `isValid` the `isLoading` state is controlled by the current page bloc, in this case, the `PaidQrcodeResumeBloc`. The `isValid` state is controlled by reusable Widget `ValueField` that has your own Bloc `ValueFieldBloc`. How I can change these properties values without `MultiBlocBuilder` ? 

Example code:
```
BlocBuilder<ValueFieldBloc, ValueFieldState>(
              builder: (context, state) {
                return ContinueButton(
                  onPressed: () => context.bloc<PaidQrcodeResumeBloc>().add(
                        PaidQrcodeResumeOnPaid(
                          widget.paidQrcode
                              .copyWith(value: moneyController.numberValue),
                        ),
                      ),
                  isValid: state is ValueFieldValid,
                  text: Strings.paid,
                  isLoading: ,
                );
              },
            )
```

**Comment by RollyPeres** [comment]
Hi @rafaelmpessoa 👋

It sounds that both `isLoading` and `isValid` should be part of the same bloc's state, `ValueFieldBloc` in your case.

You can also add an event to `ValueFieldBloc` to flip `isLoading` to true when `ContinueButton` is pressed. 

Or you could even have a `BlocListener<PaidQrcodeResumeBloc>` that would react to state changes and would add the proper event on `ValueFieldBloc` which will result in flipping `isLoading`.

Hope that gives you a couple of ideas on how you could improve your implementation 👍

**Comment by angel1st** [comment]
@felangel - I believe there is at least one good [use case over there](https://github.com/felangel/bloc/issues/538#issuecomment-541868768) and as many people insisting for the feature mentioned, there are few reasons for the request:
1. Bloc reusability - you can create a BLoC and then use it in several places. This btw is part of BLoC concept
2. BLoC maintainability - it is much easier to maintain (upgrade and test) smaller bloc with specific logic.
3. Code readability - avoiding [pyramid of doom](https://en.wikipedia.org/wiki/Pyramid_of_doom_(programming)) I believe is good programming practice in general.

I will also put here my use case since from my standing point it makes sense:

I have a calendar widget, which can be either filtered by specific appointment types (e.g. work, holiday and etc) or change its view (day, week, month and etc.).
Hence, I created two separate BLoC(s) - one for event filter `EventFilterBloc` and one for calendar view `CalendarViewBloc`. I would consider each one as a feature rather than a resource. I have a few other places with calendars, where I need `CalendarViewBloc`, but since they have different appointment types (not events), I don't need `EventFilterBloc` part there, I rather need other appointment types. Hence I created additional BLoCs - each for these different calendars and I combine them with `CalendarViewBloc`.

One more question - since we have `MultiBlocProvider` I guess some of the arguments behind creating it can apply to `MultiBlocBuilder` case, can't it?

Anyway - I will understand if the implementation is complex, but my firm belief is that it is a needed feature.

**Comment by kranfix** [comment]
Hello, I was thinking about the problem of multiple widget composition of `BlocBuilder` and `BlocListener` ,then I created an experimental package with the same API than `flutter_bloc`: https://pub.dev/packages/flutter_hooks_bloc

This package reimplements those widgets based on `HookWidget` (https://pub.dev/packages/flutter_hooks).

A widget that rebuilds depending on many bloc would be written like this:

```dart
class MyMultiBlocBuilder extends HookWidget {
  MyMultiBlocBuilder

  @override
  Widget build(BuildContext context){
    // onEmitted is called every time that the state is emitter in a cubit/bloc
    final cubitA = useBloc<CubitA, int>(onEmitted: (context, previousState, state){
      // with true, the widget rebuild, otherwise, it behave like a BlocListener
      return buildWhenA?.call(previousState, state);
    });
    final blocC = useBloc<BlocB, String>(onEmitted: (context, previousState, state){
      // If you also want to have a BlocListener behavior, you can add some code here
      if(listenWhen?.call(previousState, state) ?? true){
        listener(context, state);
      }
      return buildWhenB?.call(previousState, state);
    });
    return Column(
      children: [
        Text('${cubitA.state}'),
        Text('${blocB.state}'),
      ],
    );
  }
}
```

**Comment by kranfix** [comment]
Hello, I was thinking about the problem of multiple widget composition of `BlocBuilder` and `BlocListener` ,then I created an experimental package with the same API than `flutter_bloc`: https://pub.dev/packages/flutter_hooks_bloc

This package reimplements those widgets based on `HookWidget` (https://pub.dev/packages/flutter_hooks).

A widget that rebuilds depending on many bloc would be written like this:

```dart
class MyMultiBlocBuilder extends HookWidget {
  MyMultiBlocBuilder

  @override
  Widget build(BuildContext context){
    // onEmitted is called every time that the state is emitter in a cubit/bloc
    final cubitA = useBloc<CubitA, int>(onEmitted: (context, previousState, state){
      // with true, the widget rebuild, otherwise, it behave like a BlocListener
      return buildWhenA?.call(previousState, state);
    });
    final blocC = useBloc<BlocB, String>(onEmitted: (context, previousState, state){
      // If you also want to have a BlocListener behavior, you can add some code here
      if(listenWhen?.call(previousState, state) ?? true){
        listener(context, state);
      }
      return buildWhenB?.call(previousState, state);
    });
    return Column(
      children: [
        Text('${cubitA.state}'),
        Text('${blocB.state}'),
      ],
    );
  }
}
```

**Comment by kiesman99** [comment]
Hey @kranfix,

I think your solution looks good and would implement everything we need but there still some issues I am seeing here:

1. Following the discussion [here](https://github.com/flutter/flutter/issues/51752) and especially this [comment](https://github.com/flutter/flutter/issues/51752#issuecomment-665400546) flutter_hooks currently is **not** supposed to be an alternative state management solution. It is not ready yet and just a *POC*. Therefore, I would not use this approach as an official guide to solve this kind of problem.
2. While this solution looks slick and like something that is written very easily with n-number of needed BLoCs this might be also a Problem. As I stated in my previous comment (IMO) we do not want developers to think its good practice to use 5 BlocBuilder nested in each other. Giving a solution similar to yours we make it pretty easy to think nesting 5 BlocBuilder is looking normal. 
3. @felangel created an [issue](https://github.com/felangel/bloc/issues/712) in which he wants to discuss if he wants to keep [provider](https://pub.dev/packages/provider) as a dependency. Therefore, I don't think adding flutter_hooks as a dependency would not be the best move. You would just depend on another package. 

Maybe we should look more into [Bloc-To-Bloc-Communication](https://bloclibrary.dev/#/architecture?id=bloc-to-bloc-communication) to solve this kind of problem.

This would give the option to have one or more separate Blocs which are combined through communication. Problem here is also **combining multiple Blocs into one** would result that *including* this one *"SuperBloc*" in the UI would seem pretty normal and not very problematic to the developer.

**Comment by kiesman99** [comment]
@felangel Does it hurt performance somehow to have multiple `BlocBuilder` nested or is this something we can do without a problem?
If not my 2. argument is pretty much invalid. (If so sorry :D)

**Comment by kranfix** [comment]
@kiesman99 yeah. I know that flutter-hooks is POC, that's why I said `flutter_hooks_bloc` is an experimental package. But, it simplify many tasks that are not easy without hooks.

**Comment by felangel** [comment]
@kiesman99 nested BlocBuilders will not be less performant than hooks because at the end of the day if you want nested BlocBuilders it usually means you want something to rebuild if any of the bloc states change. In those cases, there’s nothing technically wrong with nested BlocBuilders although in my experience it’s usually a sign that the blocs are data-driven and not feature-driven. Ideally, there should be one bloc per feature in which case there should be less of a need to nest BlocBuilders. Hope that helps 👍

**Comment by angel1st** [comment]
> 
> 
> although in my experience it’s usually a sign that the blocs are data-driven and not feature-driven. Ideally, there should be one bloc per feature in which case there should be less of a need to nest BlocBuilders. Hope that helps 👍

@felangel - I believe there are at least a couple of examples e.g. [this one](https://github.com/felangel/bloc/issues/1546#issuecomment-670933264) and [this one](https://github.com/felangel/bloc/issues/538#issuecomment-669760659), where we posted our real-life examples, to show that, in more complex apps, it is not unusual to have two or even three nested `BlocBuilders` and this is not a sign of bad architectural approach.
On the other hand, if the implementation of so-called `MultiBlocBuilder` is doable even as syntactic sugar, once we do it, you will have even more happy followers 👍 
Besides - I would like also to thank you for the great library and all your efforts to maintain and develop it!

**Comment by Classy-Bear** [comment]
I'm now developing an app and faced the same problem but I conclude it was unnecessary for the following:

The definition of `BlocBuilder` according to the documentation is:

> BlocBuilder is a Flutter widget which requires a Bloc and a builder function. BlocBuilder handles building the widget in response to new states. 
[Source](https://bloclibrary.dev/#/flutterbloccoreconcepts?id=blocbuilder)

So BlocBuilder should be use only when a widget will response to new states.

<details><summary>This is wrong</summary>

 ```dart
BlocBuilder<BlocA, String>(
  builder: (context, state) {
     return Scaffold( // State is not being handle here so no need to wrap this in a BlocBuilder
      body: Text(
        state, // State is handled here
      ),
    );
  }
)
```

</details>

<details><summary>This is correct</summary>

 ```dart
Scaffold(
  body: BlocBuilder<BlocA, String>(builder: (context, state) {
    return Text(
      state, // State is handled here
    );
  }),
);
```

</details>

<details><summary>What should I do if I have a `Widget` that has multiple `Widget`s whit its own state?</summary>

```dart
Scaffold(
  body: Column(
    children: [
      BlocBuilder<BlocA, String>(builder: (context, state) {
        return Text(
          state,
         );
       }),
       BlocBuilder<BlocB, String>(builder: (context, state) {
         return Text(
          state,
        );
      }),
     ],
   ),
  );
```

</details>


<details><summary>What should I do if I have a bloc that handles the errorText of a Textfield and another bloc that handles the hintText? Use an `enum`. </summary>

```dart
Scaffold(
  body: Column(
    children: [
      BlocBuilder<BlocA, Text>(builder: (context, state) {
        return Text(
            state == Text.errorText ? 'show the error' : ''
        );
       }),
       BlocBuilder<BlocB, Text>(builder: (context, state) {
         return Text(
            state == Text.hintText ? 'show the hint' : '',
          );
      }),
    ],
  ),
);
```

</details>

Solutions to avoid this issue to be open again in the future:

* I think the BlocBuilder definition of the documentation should say something like this: _You should wrap your BlocBuilder at one level widget, don't use nested widgets_.
* Bloc Concepts best practices

People should know how to handle correctly the Bloc Concepts, that's why I opened this [issue](https://github.com/felangel/bloc/issues/1796) a week ago. This isn't part of the core functionality so doesn't add any value, it solves a problem but is not the way to do it.

Note: The code provided is for demonstration purposes.

**Comment by Deepjyoti120** [comment]
Where is BlocBuilder2 ?????? 

`BlocBuilder<UsernameupdateCubit, UsernameupdateState>(
                      builder: (context, state) {
                        return ElevatedButton(
                          style: ElevatedButton.styleFrom(
                              elevation: 1,
                              primary: Color(0xff34495e),
                              shape: RoundedRectangleBorder(
                                  borderRadius: BorderRadius.circular(14))),
                          onPressed: () {
                            if (state.userUpdateName == '') {
                              showDialog(
                                  context: context,
                                  builder: (BuildContext context) =>
                                      _NameUpdate());
                            } else {
                              lanchWhatsapp(
                                  number: "+919864044966",
                                  message: 
                                          "*Price:* \$" +
                                      "${widget.networkApi.price}");
                            }
                          },
                          child: Padding(
                            padding: const EdgeInsets.only(top: 16, bottom: 16),
                            child: Text(
                              state.userUpdateName == ''
                                  ? 'Save your Name first'
                                  : 'Order Now',
                              style: TextStyle(
                                fontWeight: FontWeight.w700,
                                // fontSize: 22,
                              ),
                            ),
                          ),
                        );
                      },
                    ),`

How to use my 2nd bloc to calculate Quantity with the current price.

because I have quantity cubit 

i want this  "${widget.networkApi.price}" to  "${widget.networkApi.price!* state.quantitystate}"
Thank you

**Comment by Deepjyoti120** [comment]
I simple use:-  

`BlocBuilder<UsernameupdateCubit, UsernameupdateState>(
                      builder: (context, state) { //Note this
                        return BlocBuilder<QuantityCubit, QuantityState>(
                          builder: (context, stateQuantity) {  //i changed variable name 
                            return ElevatedButton(`


Thank you

**Comment by felangel** [comment]
> I simply use:-
> 
> > BlocBuilder<UsernameupdateCubit, UsernameupdateState>(
> > builder: (context, state) { //Note this
> > return BlocBuilder<QuantityCubit, QuantityState>(
> > builder: (context, stateQuantity) {  //i changed variable name
> > return ElevatedButton(
> 
> Thank you

You can also use `context.watch`

```dart
@override
Widget build(BuildContext context) {
  final userState = context.watch<UsernameUpdateCubit>().state;
  final quantityState = context.watch<QuantityCubit>().state;

  return ElevatedButton(...);
}
```

**Comment by krish-bhanushali** [comment]
@felangel I know this is an old issue but I am using two cubits to manage theme and language respectively with one states each and I am passing both as individual blocbuilders over the material app. So is this a correct way to do it?  or I can use one single cubit for both.

**Comment by zombie6888** [comment]
What if instead of MultiBlocBuilder provide MultiBlocSelector in order to to combine data from different states and rebuild ui when data is changed

**Comment by felangel** [comment]
> @felangel I know this is an old issue but I am using two cubits to manage theme and language respectively with one states each and I am passing both as individual blocbuilders over the material app. So is this a correct way to do it? or I can use one single cubit for both.

I recommend one bloc/cubit per feature state.

**Comment by felangel** [comment]
> What if instead of MultiBlocBuilder provide MultiBlocSelector in order to to combine data from different states and rebuild ui when obtained data slice is changed. If you combine whole state it will work like MultiBlocBuilder

Can you provide a concrete use case in which this would be necessary? I recommend a single bloc/cubit per feature so ideally your widgets should not need to aggregate data from multiple bloc/cubit states.

**Comment by zombie6888** [comment]
It's not necessary but convenient. For example if you have have newsfeed which is contains articles and bloc for it and some other feed like favorite or something else  which is also have bloc and articles. And you have article detail page. User can go to details page from both feeds by passing article id to details route. So i want to select article from both feeds and update ui whenever article in feeds is changed. I can merge streams, use nested blocBuilders or reorganize my buisness logic (by adding third bloc, pass one bloc to another or something else) but multiselector will be more convinient for me. I think it's possible to find more use cases

**Comment by bmd007** [comment]
> 
My Dart knowledge is very low, I'm Java guy. So might be asking stupid question :))

I couldn't get this version to work. The state updates are not available to UI. I guess `watch<>()` is not as streamy as it should be.

And it apparently still needs a MultilBlocProvider anyways.

**Comment by bmd007** [comment]
> > I simply use:-
> > > BlocBuilder<UsernameupdateCubit, UsernameupdateState>(
> > > builder: (context, state) { //Note this
> > > return BlocBuilder<QuantityCubit, QuantityState>(
> > > builder: (context, stateQuantity) {  //i changed variable name
> > > return ElevatedButton(
> > 
> > 
> > Thank you
> 
> You can also use `context.watch`
> 
> ```dart
> @override
> Widget build(BuildContext context) {
>   final userState = context.watch<UsernameUpdateCubit>().state;
>   final quantityState = context.watch<QuantityCubit>().state;
> 
>   return ElevatedButton(...);
> }
> ```


My Dart knowledge is very low, I'm Java guy. So might be asking stupid question :))

I couldn't get this version to work. The state updates are not available to UI. I guess watch<>() is not as streamy as it should be.

And it apparently still needs a MultilBlocProvider anyways.

**Comment by bbura** [comment]
@felangel usecase: In the main.dart we want to globally use localisation bloc, theme bloc. I just think that nesting 2 blocbuilders is exactly like that: nasty. Anyway, good job with the bloc and also mason, great tools mate.

**Comment by narcodico** [comment]
@tananga we recommend a bloc per feature, which sometimes mean a bloc per screen if the screen covers the whole feature.

@bbura localization is UI logic and should be done in the presentation layer, not in the bloc.

**Comment by S-ecki** [comment]
I also agree with @felangel that something like a `MultiBlocBuilder` would add unnecessary complexity to Bloc and should not be part of the package. However, I needed something similar for a HomeScreen, where information from multiple blocs is read and I wanted a global loading state.

Disclaimer: In the end, I used `RxDart` to combine the state stream instead of using the `BlocBuilder3`

Still, I will share my code for the `BlocBuilder3` here, as it will save time for people who might want it in the future:


``` dart
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

typedef BlocBuilderCondition<S> = bool Function(S previous, S current);
typedef BlocWidgetBuilder3<StateA, StateB, StateC> = Widget Function(
  BuildContext,
  StateA,
  StateB,
  StateC,
);

class BlocBuilder3<
    BlocA extends StateStreamable<BlocAState>,
    BlocAState,
    BlocB extends StateStreamable<BlocBState>,
    BlocBState,
    BlocC extends StateStreamable<BlocCState>,
    BlocCState> extends StatelessWidget {

  const BlocBuilder3({
    Key? key,
    required this.builder,
    this.blocA,
    this.blocB,
    this.blocC,
    this.buildWhenA,
    this.buildWhenB,
    this.buildWhenC,
  }) : super(key: key);

  final BlocWidgetBuilder3<BlocAState, BlocBState, BlocCState> builder;

  final BlocA? blocA;
  final BlocB? blocB;
  final BlocC? blocC;

  final BlocBuilderCondition<BlocAState>? buildWhenA;
  final BlocBuilderCondition<BlocBState>? buildWhenB;
  final BlocBuilderCondition<BlocCState>? buildWhenC;

  @override
  Widget build(BuildContext context) {
    return BlocBuilder<BlocA, BlocAState>(
      bloc: blocA,
      buildWhen: buildWhenA,
      builder: (context, blocAState) {
        return BlocBuilder<BlocB, BlocBState>(
          bloc: blocB,
          buildWhen: buildWhenB,
          builder: (context, blocBState) {
            return BlocBuilder<BlocC, BlocCState>(
              bloc: blocC,
              buildWhen: buildWhenC,
              builder: (context, blocCState) {
                return builder(context, blocAState, blocBState, blocCState);
              },
            );
          },
        );
      },
    );
  }
}
```

**Comment by Livinglist** [comment]
@S-ecki how did you combine the states using RxDart? Just curious.

**Comment by S-ecki** [comment]
> @S-ecki how did you combine the states using RxDart? Just curious.

I took the state streams and used [combineLatest](https://pub.dev/documentation/rxdart/latest/rx/CombineLatestStream-class.html) if I remember correctly. 
Note that with `combineLatest` only emits after each stream emitted an event by default.

**Comment by amirbahrawy** [comment]
> > I simply use:-
> > > BlocBuilder<UsernameupdateCubit, UsernameupdateState>(
> > > builder: (context, state) { //Note this
> > > return BlocBuilder<QuantityCubit, QuantityState>(
> > > builder: (context, stateQuantity) {  //i changed variable name
> > > return ElevatedButton(
> > 
> > 
> > Thank you
> 
> You can also use `context.watch`
> 
> ```dart
> @override
> Widget build(BuildContext context) {
>   final userState = context.watch<UsernameUpdateCubit>().state;
>   final quantityState = context.watch<QuantityCubit>().state;
> 
>   return ElevatedButton(...);
> }
> ```

Thats awesome i used the same approach also you can use  select if you need to listen on only value like this 
final userState = context.select((UsernameUpdateCubit cubit) => cubit.state);

**Comment by dittombek** [comment]
Please take a look of the doc flutter bloc [https://bloclibrary.dev/#/flutterbloccoreconcepts?id=usage-2](url)

context.watch must inside of builder

`@override
Widget build(BuildContext context) {
  return Builder(
     builder: (context) {
        final stateA = context.watch<BlocA>().state;
        final stateB = context.watch<BlocB>().state;
        final stateC = context.watch<BlocC>().state;

        return Text('Hello World');
     }
  );
}
`

**Comment by dittombek** [comment]
Please take a look of this flutter bloc documents [https://bloclibrary.dev/#/flutterbloccoreconcepts?id=usage-2](url)

To implements multiple blocbuilder, use Builder and context.watch.

`@override
Widget build(BuildContext context) {
  return Builder(
    builder: (context) {
      final stateA = context.watch<BlocA>().state;
      final stateB = context.watch<BlocB>().state;
      final stateC = context.watch<BlocC>().state;

      return const Text('Hello World');
    },
  );
}`

**Comment by dittombek** [comment]
Please take a look of this flutter bloc documents [https://bloclibrary.dev/#/flutterbloccoreconcepts?id=usage-2](url)

To implements multiple blocbuilder, use Builder and context.watch.

`
@override
Widget build(BuildContext context) {
  return Builder(
    builder: (context) {
      final stateA = context.watch<BlocA>().state;
      final stateB = context.watch<BlocB>().state;
      final stateC = context.watch<BlocC>().state;

      return const Text('Hello World');
    },
  );
}
`

**Comment by dittombek** [comment]
Please take a look of this flutter bloc documents [https://bloclibrary.dev/#/flutterbloccoreconcepts?id=usage-2](url)

For safe implementation of multiple blocbuilder, use Builder and context.watch.

```
@override
Widget build(BuildContext context) {
  return Builder(
    builder: (context) {
      final stateA = context.watch<BlocA>().state;
      final stateB = context.watch<BlocB>().state;
      final stateC = context.watch<BlocC>().state;

      return const Text('Hello World');
    },
  );
}
```

**Comment by SantiiRepair** [comment]
`MultiBlocBuilder` would be interesting, `BlocBuilder2`... doesn't look good, I would like to know why `MultiBlocBuilder` is not a good idea

**Comment by Fingertips18** [comment]
For clean architecture purposes, `MultiBlocBuilder`, `MultiBlocListener` or `MultiBlocConsumer`  would be useful. I encountered a situation where I wanted to add two global blocs on top of a particular feature (which also has a bloc itself). Here is the idea behind what I was trying to implement:

```
BlocBuilder<Global1Bloc, Global1State>(
      buildWhen: (previous, current) {
        return previous.loading != current.loading;
      },
      builder: (context, global1State) {
        return BlocConsumer<Global2Bloc, Global2State>(
          listenWhen: (previous, current) {
            return !current.loading;
          },
          listener: (context, state) {
            if(state.isSuccess) {...};
          },
          buildWhen: (previous, current) {
            return previous.loading != current.loading;
          },
          builder: (context, global2State) {
            return BlocBuilder<FeatureBloc, FeatureState>(
              buildWhen: (previous, current) {
                return previous.hasInput!= current.hasInput;
              },
              builder: (context, featureState) {
                return CustomTextButton(
                  bgColor: Colors.black,
                  label: "Confirm",
                  loading: firebaseState.loading || userState.loading,
                  onPressed: {...},
                );
              },
            );
          },
        );
      },
    );
  }
```

This is working but its kinda messy to look at. I will stick to this pattern for the meantime but might follow https://github.com/felangel/bloc/issues/538#issuecomment-1711049615

**Comment by zombie6888** [comment]
> For clean architecture purposes, `MultiBlocBuilder`, `MultiBlocListener` or `MultiBlocConsumer` would be useful. I encountered a situation where I wanted to add two global blocs on top of a particular feature (which also has a bloc itself). Here is the idea behind what I was trying to implement:
> 
> ```
> BlocBuilder<Global1Bloc, Global1State>(
>       buildWhen: (previous, current) {
>         return previous.loading != current.loading;
>       },
>       builder: (context, global1State) {
>         return BlocConsumer<Global2Bloc, Global2State>(
>           listenWhen: (previous, current) {
>             return !current.loading;
>           },
>           listener: (context, state) {
>             if(state.isSuccess) {...};
>           },
>           buildWhen: (previous, current) {
>             return previous.loading != current.loading;
>           },
>           builder: (context, global2State) {
>             return BlocBuilder<FeatureBloc, FeatureState>(
>               buildWhen: (previous, current) {
>                 return previous.hasInput!= current.hasInput;
>               },
>               builder: (context, featureState) {
>                 return CustomTextButton(
>                   bgColor: Colors.black,
>                   label: "Confirm",
>                   loading: featureState.loading || global1State.loading,
>                   onPressed: {...},
>                 );
>               },
>             );
>           },
>         );
>       },
>     );
>   }
> ```
> 
> This is working but its kinda messy to look at. I will stick to this pattern for the meantime but might follow [#538 (comment)](https://github.com/felangel/bloc/issues/538#issuecomment-1711049615)

Instead of buildwhen and listenWhen use selectors. You don't need blocBuilders here at all. You will be surprised how clean your code will become. It's not realted to clean architechture beacuse it's all about presentation layer
